dirty_cat/_deduplicate.py

```diff
@@ -0,0 +1,175 @@
+"""
+Implements deduplication based on clustering string distance matrices.
+This works best if there is a number of underlying categories that
+sometimes appear in the data with small variations and/or misspellings.
+"""
+from typing import List, Literal, Optional, Sequence, Tuple, Union
+
+import numpy as np
+import pandas as pd
+from scipy.cluster.hierarchy import fcluster, linkage
+from scipy.spatial.distance import pdist, squareform
+from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
+from sklearn.metrics import silhouette_score
+
+
+def compute_ngram_distance(
+    unique_words: Union[Sequence[str], np.ndarray],
+    ngram_range: Tuple[int, int] = (2, 4),
+    analyzer: str = "char_wb",
+) -> np.ndarray:
+    """Computes the condensed n-gram distance matrix between words in
+    `unique_words`, using the
+    term frequency-inverse document frequency (tf-idf).
+
+    Parameters
+    ----------
+    unique_words : Union[Sequence[str], np.ndarray]
+        Sequence or array of unique words from the original data.
+    ngram_range : Tuple[int, int], optional
+        The n-gram range to compute the distance in, by default (2, 4)
+    analyzer : str, optional
+        Analyzer to extract n-grams, by default "char_wb"
+
+    Returns
+    -------
+    np.ndarray
+        An n-times-(n-1)/2 matrix of n-gram TfIdf distances between `unique_words`.
+    """
+    enc = CountVectorizer(ngram_range=ngram_range, analyzer=analyzer)
+    encoded = TfidfTransformer().fit_transform(enc.fit_transform(unique_words))
+
+    distance_mat = pdist(encoded.todense(), metric="euclidean")
+    return distance_mat
+
+
+def deduplicate(
+    data: Sequence[str],
+    n_clusters: Optional[int] = None,
+    ngram_range: Tuple[int, int] = (2, 4),
+    analyzer: Literal["word", "char", "char_wb"] = "char_wb",
+    method: Literal[
+        "single", "complete", "average", "centroid", "median", "ward"
+    ] = "average",
+    return_translation_table: bool = False,
+) -> Union[List[str], Tuple[List[str], pd.Series]]:
+    """Deduplicates data by computing the n-gram distance between unique
+    categories in data, performing hierarchical clustering on this distance
+    matrix, and choosing the most frequent element in each cluster as the
+    'correct' spelling. This method works best if the true number of
+    categories is significantly smaller than the number of observed spellings.
+
+    Parameters
+    ----------
+    data : Sequence[str]
+        The data to be deduplicated.
+    n_clusters : Optional[int], optional
+        number of clusters to use for hierarchical clustering, if None use the
+        number of clusters that lead to the lowest silhouette score,
+        by default None
```

Comment:
On the docs, best is to see what was used before ([e.g. fuzzy_join](https://github.com/dirty-cat/dirty_cat/blob/master/dirty_cat/_fuzzy_join.py)) and try to implement the same. 
We try to stay consistent with work on previous features. For instance, this should be:
```suggestion
    n_clusters : typing.Optional[int], default=None
        Number of clusters to use for hierarchical clustering, if None use the
        number of clusters that lead to the lowest silhouette score.
```